Add historical release notes through 0.30.1

mDuo13 opened the pull request:
This adds a "ReleaseNotes.md" file with pretty much the entire [wiki release notes category](https://wiki.ripple.com/Category:Rippled_release_notes) ported to Markdown. The plan being any new release can have its release notes added to the top of the file.

Not included:
- releases prior to 0.12.0 (no release notes for those versions are known)
- version 0.13.0 (release notes appear not to have been finished and the tag for this release doesn't exist)
- release notes for 0.31.0 (drafted, but not included yet)

nbougalis commented:
Some issues:

First, if we include the release notes in the repo, we cannot include the blurb that has the commit ID of the version-setting commit - you can't know that until after the commit is checked in and it depends on the contents of the commit itself.

The bigger question is why do we need to _entire_ history of release notes? It makes more sense to have the file only include the nodes for the latest release. Notes for previous releases can be found easily enough by going back to that tag and checking them out.

The solution that I would propose is to include the release notes for the new version packaged with the version-setting commit and simply not list the expected tip commit ID, but to simply include a link to the tag (e.g. https://github.com/ripple/rippled/tree/0.30.1)

@yana-novikova, @vinniefalco: thoughts?

vinniefalco commented:
If all you have is a working set, you have no way to know the changes by release if only the latest release' notes are present.

mDuo13 commented:
@nbougalis For future release notes, my plan was to do pretty much exactly as you described (leave out the bit about which commit should be at the tip of the release, but leave in a link to the tag).

The main reason I want to include the entire history of release notes is that for any tag that exists today, you _can't_ do as you suggested and get the "latest" release notes by that tag -- there's nothing there today and we can't add them retroactively. When wiki.ripple.com eventually goes away, the old release notes need to live somewhere. May as well just have them in the file. It's also convenient if all you have is a tarball or something. No need to look elsewhere for all the release notes. I think it really is OK to have all the notes in one file, at least until the file starts getting to >1MB territory.

I do also want to edit the comments related to the ["Release" tags on GitHub](https://github.com/ripple/rippled/releases/), but I think someone with higher access privileges than me will have to do that. The good news is you _can_ retroactively edit those, as I've [demonstrated in my repo](https://github.com/mDuo13/rippled/releases/tag/0.25.1).

vinniefalco commented:
I agree with @mDuo13 (I've added you as a collaborator so the tag comments are editable).

My only complaint with the release notes is that the fonts are enormous (they are using the markdown heading styles) and its hard to tell where one release note ends and the other begins. I don't know what the answer to it is, maybe a horizontal rule or something. I would prefer if there was only one point size (say, 12 or 14) and the use only of the bold style to indicate section headings within an individual release note.

mDuo13 commented:
Revised as @vinniefalco suggested: converted the sub-headers to just bold text, and added horizontal rules to divide the release notes.

I left the release note titles as h1s because that's the semantically correct hierarchy and the large font also makes it easier to visually distinguish one release from another.

vinniefalco commented:
Enormous improvement with the latest change :+1:

nbougalis commented:
Eh, I'm not sold on having the entire release note history in a file, but I'm not _too_ worried about it. I think the ideal situation would be to include the changes to the release notes in the version-setting commit (and in the commit message itself) instead of as a separate commit.

We can then use the GitHub "Release" tab well.

vinniefalco commented:
@nbougalis Merging the release notes into the commit that sets the version would lose commit authorship information.

yana-novikova commented:
should we include the list of all release notes on top of the file and make them hyperlinks that you can click to go to the particular release?

vinniefalco commented:
@yana-novikova Hmm... I don't think so, that list would get bigger and bigger. Users can just search or scroll. Its not something that happens so often that we need to facilitate it.

nbougalis commented:
@vinniefalco: does it really matter who authored the release notes - and is it always one person anyways? It makes more sense (to me) to have the update to the release notes in the version-setting commit - that makes it trivial for someone to look at the release notes associated with the particular version, especially since the tag itself points to that commit.

nbougalis commented:
Merged as 3f84dd8cf93ca1cbaef24164edf393a664e72ab6
